### PR TITLE
Optimize OnPersists

### PR DESCRIPTION
**Pull request by shargon**
Close https://github.com/neo-project/neo/issues/3145

### CODE REVIEW

File: src/Neo/SmartContract/Native/NeoToken.cs [hunk 1]
```diff
@@ -226,6 +227,7 @@ internal override ContractTask Initialize(ApplicationEngine engine)
             engine.Snapshot.Add(CreateStorageKey(Prefix_Committee), new StorageItem(cachedCommittee));
             engine.Snapshot.Add(CreateStorageKey(Prefix_VotersCount), new StorageItem(System.Array.Empty<byte>()));
             engine.Snapshot.Add(CreateStorageKey(Prefix_GasPerBlock).AddBigEndian(0u), new StorageItem(5 * GAS.Factor));
+            engine.Snapshot.Add(CreateStorageKey(Prefix_GasPerBlockCurrent), new StorageItem(new GasPerBlock() { Gas = 5 * GAS.Factor, PreviousGas = 5 * GAS.Factor, Index = 0 }));
```
Comment (cschuchardt88) [inline src/Neo/SmartContract/Native/NeoToken.cs:230]:
```suggestion
            engine.Snapshot.Add(CreateStorageKey(Prefix_GasPerBlockCurrent).AddBigEndian(0u), new StorageItem(5 * GAS.Factor));
```

This way you can use `backwards` for lookup on the index

**Comment by Jim8y** [comment]
A glance of this pr, no idea what its for. Would be great to have some comments in the code, and description in the pr.

File: src/Neo/SmartContract/Native/NeoToken.cs (hunk 1 above)
Comment (shargon) [reply src/Neo/SmartContract/Native/NeoToken.cs:230]: This pr is for avoid `backwards`

**Comment by shargon** [comment]
> A glance of this pr, no idea what its for. Would be great to have some comments in the code, and description in the pr.

https://github.com/neo-project/neo/issues/3145

File: src/Neo/SmartContract/Native/NeoToken.cs [hunk 2]
```diff
@@ -283,9 +285,18 @@ private void SetGasPerBlock(ApplicationEngine engine, BigInteger gasPerBlock)
                 throw new ArgumentOutOfRangeException(nameof(gasPerBlock));
             if (!CheckCommittee(engine)) throw new InvalidOperationException();
 
-            uint index = engine.PersistingBlock.Index + 1;
-            StorageItem entry = engine.Snapshot.GetAndChange(CreateStorageKey(Prefix_GasPerBlock).AddBigEndian(index), () => new StorageItem(gasPerBlock));
+            var gp = new GasPerBlock() { Gas = gasPerBlock, Index = engine.PersistingBlock.Index + 1 };
```
Comment (vncoelho) [inline src/Neo/SmartContract/Native/NeoToken.cs:288]: one argument is missing here, no?
The correct constructor requires 3

Comment (shargon) [reply src/Neo/SmartContract/Native/NeoToken.cs:267]: `PreviousGas` default value is 0

File: src/Neo/SmartContract/Native/NeoToken.cs [hunk 3]
```diff
@@ -620,6 +642,26 @@ protected override StackItem ElementToStackItem((ECPoint PublicKey, BigInteger V
...
+            public uint Index;
+            public BigInteger Gas;
+            public BigInteger PreviousGas;
+
+            public void FromStackItem(StackItem stackItem)
+            {
+                Struct @struct = (Struct)stackItem;
+                Index = (uint)@struct[0].GetInteger();
+                Gas = @struct[1].GetInteger();
+                Gas = @struct[2].GetInteger();
```
Comment (shargon) [inline src/Neo/SmartContract/Native/NeoToken.cs:656]:
```suggestion
                PreviousGas = @struct[2].GetInteger();
```

File: src/Neo/SmartContract/Native/NeoToken.cs [hunk 4]
```diff
@@ -184,6 +185,7 @@ internal override ContractTask InitializeAsync(ApplicationEngine engine, Hardfor
                 engine.Snapshot.Add(CreateStorageKey(Prefix_Committee), new StorageItem(cachedCommittee));
                 engine.Snapshot.Add(CreateStorageKey(Prefix_VotersCount), new StorageItem(System.Array.Empty<byte>()));
                 engine.Snapshot.Add(CreateStorageKey(Prefix_GasPerBlock).AddBigEndian(0u), new StorageItem(5 * GAS.Factor));
+                engine.Snapshot.Add(CreateStorageKey(Prefix_GasPerBlockCurrent), new StorageItem(new GasPerBlock() { Gas = 5 * GAS.Factor, PreviousGas = 5 * GAS.Factor, Index = 0 }));
```
Comment (roman-khimov) [inline src/Neo/SmartContract/Native/NeoToken.cs:188]: Please, do not change the storage scheme, it's very easy to cache the value.

**Comment by shargon** [comment]
> 1. We need some number to proof this is worth doing at all.
> 2. How often people request this via RPC?

is not only requested by rpc, also is used on each block

**Comment by roman-khimov** [comment]
This returns us to #3145. It's very easy to cache for blocks (and not only for blocks). Storage scheme change that duplicates some already existing data should have more justification to me.

**Comment by Jim8y** [comment]
@neo-project/core what about this pr?

**Comment by shargon** [comment]
I will use a different approach

**Pull request closed without merging**